### ISSUE TITLE
Enable performance monitoring using Elastic Network Adapter (ENA)

### ISSUE THREAD

**Issue by ctgardner** [1debit/alternat#74]
tl;dr - Can alterNAT instances publish additional network metrics to CloudWatch?

By default, AWS CloudWatch provides just four network metrics for EC2 instances: `Network{In,Out}` and `NetworkPackets{In,Out}`. In particular, there's no metric indicating the available network credits or whether an EC2 instance is being throttled.† (Applicable to instances with burstable network performance only.)

The good news is [Elastic Network Adapter](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/monitoring-network-performance-ena.html) (ENA) provides addition network metrics which can be published to CloudWatch using the [CloudWatch agent](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/install-CloudWatch-Agent-on-EC2-Instance-fleet.html#start-CloudWatch-Agent-EC2-fleet). Some useful metrics include:

Metric | Description | Supported on
-- | -- | --
bw_in_allowance_exceeded | The number of packets queued or dropped because the inbound aggregate bandwidth exceeded the maximum for the instance. | All instance types
bw_out_allowance_exceeded | The number of packets queued or dropped because the outbound aggregate bandwidth exceeded the maximum for the instance. | All instance types
pps_allowance_exceeded | The number of packets queued or dropped because the bidirectional PPS exceeded the maximum for the instance. | All instance types

These metrics indicate when an alterNAT instance's network performance is throttled, which is very useful to know. Alarms can be created and configured to send notifications when this occurs.

To publish these metrics, some relatively simple changes are required to an already deployed alterNAT setup (written in [CDKTF](https://developer.hashicorp.com/terraform/cdktf)).

Firstly, an inline policy is required to allow the instance to publish metrics to CloudWatch:

```ts
const alterNat = new alternat.Alternat(this, "alternat", {
  additionalInstancePolicies: [
    {
      policy_name: "publish-cloudwatch-custom-metrics",
      policy_json: Fn.jsonencode({
        Version: "2012-10-17",
        Statement: [
          {
            Action: "cloudwatch:PutMetricData",
            Effect: "Allow",
            Resource: "*",
          },
        ],
      }),
    },
  ],
  // ...
})
```

And an user data script is required to install and configure the CloudWatch agent:

```ts
const userdata = new TerraformAsset(this, "userdata", {
  path: "src/userdata/install-cw-agent.sh",
  type: AssetType.FILE,
})

const alterNat = new alternat.Alternat(this, "alternat", {
  // ...
  natInstanceUserDataPostInstall: Fn.file(userdata.path),
  // ...
})
```

```sh
#!/bin/bash

CW_AGENT_CONFIG_PATH='/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json'

sudo yum -y install amazon-cloudwatch-agent

sudo touch "$CW_AGENT_CONFIG_PATH"
sudo tee "$CW_AGENT_CONFIG_PATH" > /dev/null << EOF
{
  "metrics": {
    "append_dimensions": {
      "InstanceId": "\${aws:InstanceId}"
    },
    "metrics_collected": {
      "ethtool": {
        "metrics_include": [
          "bw_in_allowance_exceeded",
          "bw_out_allowance_exceeded",
          "pps_allowance_exceeded"
        ]
      }
    }
  }
}
EOF

# load config file and restart agent
sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl \
  -a fetch-config \
  -m ec2 \
  -s \
  -c "file:$CW_AGENT_CONFIG_PATH"
```

That's it! Now, enhanced network metrics are published to CloudWatch by each alterNAT instance. As mentioned, these metrics are useful for indicating when an instance is throttled because the (hidden) network credits have been exhausted. They can also be used to optimise the instance size, as you can test different sizes and you'll know when an instance has insufficient network performance.

I'm sharing this because I think it's useful and I'd like others to benefit. Could this feature be included with alterNAT?

_† I've contacted AWS Premium Support and they've confirmed they're actively working on a feature which exposes this information._

**Comment by bwhaley** [comment]
We could certainly do this. Contributions are most welcome, so feel free to submit a pull request if you have the time!

**Comment by bwhaley** [comment]
Closing this since it hasn't been active in a while. Do feel free to submit a PR if you'd like to see this enhancement!
